packaging/userspace/runpath_util.py

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python3
+
+import shutil, sys, os, stat
+import logging
+import errno
+import argparse
+from os import path
+from concurrent.futures import Future
+
+import script_common
+from context import PackagerContext
+from utilbase import PackagerUtilBase
+
+__all__ = ['RunpathUtil']
+
+class RunpathUtil(PackagerUtilBase):
+
+	def __init__(self, context, executor=None, chrpath_path=None):
+		super().__init__(context, executor)
+		self.chrpath_path = chrpath_path
+
+	def _set_rpath_chrpath(self, bin_paths, runpath, chrpath_path, env):
+		l = logging.getLogger(__name__)
+		ilib = self.ilib
+
+		env = self.fill_out_envdict(env)
+
+		for bin_path, output_path in bin_paths:
+			if path.abspath(bin_path) != path.abspath(output_path):
+				shutil.copyfile(bin_path, output_path)
+
+		bin_paths = [output_path for bin_path, output_path in bin_paths]
+
+		chrpath_args = [chrpath_path, '-c', '-r', runpath]
+		chrpath_args.extend(bin_paths)
+
+		ilib.execute_command(chrpath_args, env=env)
+
+		for bin_path in bin_paths:
+			os.chmod(bin_path, stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR |
+			                   stat.S_IRGRP |                stat.S_IXGRP |
+			                   stat.S_IROTH |                stat.S_IXOTH)
+
+		return bin_paths
+
+	def _set_rpath_lief(self, bin_path, runpath, set_origin, output_path):
+		l = logging.getLogger(__name__)
+		lief = self.lief
+
+		l.info(bin_path + ': Setting runpath')
+
+		# lief tends to segfault with libc++. short-circuit back to chrpath
+		fname = path.basename(bin_path)
+		if fname.startswith('libc++.') or fname.startswith('libc++abi.'):
+			# TODO: we need env from the caller
+			env = self.fill_out_envdict()
+			chrpath_path = self.find_tool(['chrpath'], self.chrpath_path, raise_if_not_found=False, env=env)
+			# TODO: callers are not prepared to handle a None response here.
+			if not chrpath_path:
+				return None
+			return self._set_rpath_chrpath([(bin_path, output_path)], runpath, chrpath_path, env)
+
+		lib = lief.parse(bin_path)
+
+		if set_origin:
+			if lib.has(lief.ELF.DYNAMIC_TAGS.FLAGS):
+				ent_flags = lib.get(lief.ELF.DYNAMIC_TAGS.FLAGS)
+				ent_flags.add(lief.ELF.DYNAMIC_FLAGS.ORIGIN)
+			else:
+				ent_flags = lief.ELF.DynamicEntryFlags(lief.ELF.DYNAMIC_TAGS.FLAGS,
+					int(lief.ELF.DYNAMIC_FLAGS.ORIGIN))
+				lib.add(ent_flags)
+
+			if lib.has(lief.ELF.DYNAMIC_TAGS.FLAGS_1):
+				ent_flags_1 = lib.get(lief.ELF.DYNAMIC_TAGS.FLAGS_1)
+				ent_flags_1.add(lief.ELF.DYNAMIC_FLAGS_1.ORIGIN)
+			else:
+				ent_flags_1 = lief.ELF.DynamicEntryFlags(lief.ELF.DYNAMIC_TAGS.FLAGS_1,
+					int(lief.ELF.DYNAMIC_FLAGS_1.ORIGIN))
+				lib.add(ent_flags_1)
+
+		# if we have rpath but no runpath, change rpath to runpath
+		if lib.has(lief.ELF.DYNAMIC_TAGS.RPATH) and not lib.has(lief.ELF.DYNAMIC_TAGS.RUNPATH):
+			lib.get(lief.ELF.DYNAMIC_TAGS.RPATH).tag = lief.ELF.DYNAMIC_TAGS.RUNPATH
```

Comment:
just a check ... is this backwards?   a quick read expected it the other way around.

Reply:
We prefer `RUNPATH` over `RPATH`, so unless I'm missing something, this is correct.

Reply:
i see it now - definitely read it backwards myself.